Support regex delimiter and limit argument for Spark split function

gaoyangxiaozhu opened the pull request:
Fixes Spark split function.
1) Supports splitting with a general regex pattern.
2) Supports `limit` parameter.
3) Rewrites as simple function.

Issue: https://github.com/facebookincubator/velox/issues/4066
Spark document: https://spark.apache.org/docs/latest/api/sql/#split
Spark implementation: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L523-L552

gaoyangxiaozhu commented:
@rui-mo / @PHILO-HE for help review.

rui-mo commented:
@gaoyangxiaozhu Thanks for your work. There is an existing PR supporting regex in split function https://github.com/facebookincubator/velox/pull/6155, and maybe we need to confirm its status first.

gaoyangxiaozhu commented:
> @gaoyangxiaozhu Thanks for your work. There is an existing PR supporting regex in split function #6155, and maybe we need to confirm its status first. cc: @PHILO-HE

oh, i see, i didn't go through the PRs to check if already people have touch this work. sorry.

gaoyangxiaozhu commented:
> @gaoyangxiaozhu Thanks for your work. There is an existing PR supporting regex in split function #6155, and maybe we need to confirm its status first. cc: @PHILO-HE

btw, @rui-mo / @PHILO-HE . looks https://github.com/facebookincubator/velox/pull/6155 only make support for `constant` pattern, not cover for non-constant pattern or limit parameter. Will confirm in the PR comments

gaoyangxiaozhu commented:
@rui-mo / @PHILO-HE / @FelixYBW since the PR https://github.com/facebookincubator/velox/pull/6155 only consider constant pattern support and has no update long time, could you start review my Pr for `split` improvement to let's help make this part finalize asap ?

PHILO-HE commented:
> @rui-mo / @PHILO-HE / @FelixYBW since the PR #6155 only consider constant pattern support and has no update long time, could you start review my Pr for `split` improvement to let's help make this part finalize asap ?

@jackylee-ch, what do you think? I guess you took over the work of 6155.

jackylee-ch commented:
> @rui-mo / @PHILO-HE / @FelixYBW since the PR #6155 only consider constant pattern support and has no update long time, could you start review my Pr for `split` improvement to let's help make this part finalize asap ?

Sorry for late response. It's ok to move on this PR since it covers #6155. Just concern about the performance impact of this PR. @gaoyangxiaozhu, maybe create a benchmark in gluten and make sure it won't cause negative performance gains?

rui-mo reviewed (commented):
Could you check the corner cases and corresponding tests from my PR first? Thanks.
https://github.com/facebookincubator/velox/pull/8825/files#diff-ebc054da2ecc11ac3cf20d47cbc112f9c3b9cf3d5f1784b2149fb0f4e997a2d8R127-R132

rui-mo commented on velox/docs/functions/spark/string.rst:
```diff
@@ -197,7 +197,8 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
 
-    Splits ``string`` on ``delimiter`` and returns an array. ::
+    Splits ``string`` on ``delimiter`` and returns an array.
+    The delimiter is any string matching regex, supported by re2. ::
```
Please also update the documentation for limit if it is added. Some draft: https://github.com/facebookincubator/velox/pull/8825/files#diff-945c1e481b55ff702e669348c241946c1b7851d575913d1a0da54952750f96daR193-R199.

gaoyangxiaozhu replied:
updated, thanks

rui-mo commented on velox/functions/sparksql/tests/SplitFunctionsTest.cpp:
```diff
@@ -14,96 +14,350 @@
+  /// function arguments.
+  /// limit should be set to the corresponding limit, if query contains limit
+  /// argument (C2).
+  VectorPtr run(
```
Function `testEncodings` could be used to test different encodings.

gaoyangxiaozhu replied:
the `testEncondings` is not good enough for our tests since it only test with inputs all has one encoding , for example the inputs all are `flat`, or `dictionary` or `constant`. 

But our code path also need verify for each input has different encoding as `string` is `flat` encoding, others (delim, pattern) is `constant`  etc to verify two code path
https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/functions/sparksql/SplitFunctions.cpp#L52C1-L53C52 and https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/functions/sparksql/SplitFunctions.cpp#L52C1-L53C52

This part also follow similar way of how presto `splitTest` does - https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/functions/prestosql/tests/SplitTest.cpp#L42C1-L45C39

rui-mo replied:
@gaoyangxiaozhu Shall we extend the testEncodings API by adding a testMixed function which generates mixed encodings and verify the result? I'm glad to support that in a separate PR.

gaoyangxiaozhu commented:
> @gaoyangxiaozhu Could you check the corner cases and corresponding tests from my PR first? Thanks. https://github.com/facebookincubator/velox/pull/8825/files#diff-ebc054da2ecc11ac3cf20d47cbc112f9c3b9cf3d5f1784b2149fb0f4e997a2d8R127-R132

updated and added ut to cover - https://github.com/gayangya/velox/blob/eddfb43831ac083a53ad17fcdca28bbd0c4b1e37/velox/functions/sparksql/tests/SplitFunctionsTest.cpp#L250

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,189 @@
+        const auto delim = delims->valueAt<StringView>(row);
+        if (delim.size() == 0) {
+          splitEmptyDelimer(
+              strings->valueAt<StringView>(row), limit, row, resultWriter);
```
I assume limit could be non-constant, and we shall get it from the limits vector.

gaoyangxiaozhu replied:
good catch, thank you @rui-mo

PHILO-HE commented on velox/docs/functions/spark/string.rst:
```diff
@@ -195,18 +195,27 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+        SELECT split('abcd', ''); -- ["a", "b", "c", "d"]
+        SELECT split('abcd', '', 3); -- ["a", "b", "c"]
 
 .. spark:function:: split(string, delimiter, limit) -> array(string)
```
Looks this has been covered by the above. Maybe, we can remove it.

gaoyangxiaozhu replied:
the above one `SELECT split('one', ''); -- ["o", "n", "e", ""]` it not right, i will remove it.

PHILO-HE commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
-class Split final : public exec::VectorFunction {
- public:
-  Split() {}
+  mutable functions::detail::ReCache cache_;
```
Better to put variable after method member.

PHILO-HE commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
+      rows.applyToSelected([&](vector_size_t row) {
+        const auto delim = delims->valueAt<StringView>(row);
+        if (!noLimit) {
+          limit = limits->valueAt<int64_t>(row);
```
Consider two cases: 1) delim is constant, limit is not; 2) limit is constant, delim is not. Seems they also go into this else branch, but causes issues in using `->valueAt<>(row)`?

gaoyangxiaozhu replied:
yes, it would go to `general` else path but it shouldn't cauases any issues here right ? It just means we don't do optimization for `flat, const, *` or `flat, *, const` cases, i think that should be fine, @PHILO-HE , we can do refactor in following PR if we find that change is needed for perf purpose.

PHILO-HE reviewed (commented):
Please revise title/description, as refactor is not the primary work of this pr. Maybe, "Support regex delimiter and limit argument for Spark split function"?

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
-  char charPattern = pattern.data()[0];
-  // TODO: Add support for zero-length pattern, 2-character pattern
-  // TODO: add support for general regex pattern using R2
-  return std::make_shared<SplitCharacter>(charPattern);
```
Please add some checks for the input argument size and types.
https://github.com/facebookincubator/velox/pull/8825/files#diff-ebc054da2ecc11ac3cf20d47cbc112f9c3b9cf3d5f1784b2149fb0f4e997a2d8R204-R216

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // adding them to the elements vector, until we reached the end of the
+    // string or the limit.
```
adding -> add, reached -> reach

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
+    // adding them to the elements vector, until we reached the end of the
+    // string or the limit.
+    int32_t addedElements{0};
+    bool emptyDelim = delim.size() == 0 ? true : false;
```
It seems this variable is not in use. Could you confirm?

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
+        re2String, pos, input.size(), RE2::Anchor::UNANCHORED, subMatches, 1)) {
+      const auto fullMatch = subMatches[0];
+      auto offset = fullMatch.data() - start;
+      const auto size = fullMatch.size();
```
nit: it seems this variable is used only once. Can we remove it?

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
+    }
+
+    // Add the rest of the string and we are done.
+    // Note, that the rest of the string can be empty - we still add it.
```
nit: remove the `,` after note.

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,195 @@
+    resultWriter.commit();
   }
+
+  mutable functions::detail::ReCache cache_;
```
nit: add comments for this variable.
I wonder why we need the keyword mutable. It allows us to modify this variable from const functions, but in current implementation cache_  has not been modified. Am I miss something?

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
-/// This class only implements the basic split version in which the pattern is a
-/// single character
-class SplitCharacter final : public exec::VectorFunction {
+/**
```
nit: for multiple-line comment, we need to use `///` for public APIs or `//` otherwise. Here `//` is needed as this class is in an anonymous namespace, seeing https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md#comments.
Comment suggestion:
>// split(string, delimiter[, limit]) -> array(varchar)
//
// Splits string on delimiter and returns an array of size at most limit.
// Delimiter is a string representing regular expression. Limit is an integer
// which controls the number of times the regex is applied. By default, limit is
// -1.

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
+ * limit is an integer which controls the number of times the regex is applied.
+ * By default, limit is -1.
+ *
+ * If delimiter is not empty, then when limit > 0, the last element in
```
Perhaps move the detailed description for `delimiter is not empty` to the splitInner function.

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
-    SplitCharacter splitCharacter(pattern);
-    splitCharacter.apply(rows, args, nullptr, context, result);
+  // Split with a non-empty pattern.
+  void splitInner(
```
Comment suggestion:
>   // Split with a non-empty delimiter. If limit > 0, The resulting array's
  // length will not be more than `limit` and the resulting array's last entry
  // will contain all input beyond the last matched regex. If limit <= 0,
  // delimiter will be applied as many times as possible, and the resulting
  // array can be of any size.

nit: perhaps rename this function as `split` corresponding with splitEmptyDelimiter.

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
-
-      resultWriter.commit();
-    });
+    // Optimization for the (flat, const, const) case.
```
nit: Fast path for (flat, const, const).

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
+      const auto* rawStrings = strings->data<StringView>();
+      const auto delim = delims->valueAt<StringView>(0);
+      int32_t limit = std::numeric_limits<int32_t>::max();
+      bool isEmptyDelimiter = delim.empty();
```
nit: to improve readability, move this line after L73 because this part is deciding the limit.

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
+          limit = constant;
+        }
+      }
+      rows.applyToSelected([&](vector_size_t row) {
```
isEmptyDelimiter is constant, we don't need to check it for each row.
```
      if (isEmptyDelimiter) {
        rows.applyToSelected([&](vector_size_t row) {
          splitEmptyDelimiter(rawStrings[row], limit, row, resultWriter);
        });
      } else {
        rows.applyToSelected([&](vector_size_t row) {
          splitInner(rawStrings[row], delim, limit, row, resultWriter);
        }); 
      }
```

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
+      });
+    } else {
+      // The rest of the cases are handled through this general path and no
+      // direct access.
```
nit: this comment could be removed as it does not contain detailed information.

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
+            limit = limitValue;
+          }
+        }
+        if (delim.size() == 0) {
```
nit: delim.empty()

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    VELOX_CHECK(!delim.empty(), "Delimiter size should not empty");
```
Use VELOX_USER_CHECK instead. The throw of VELOX_CHECK indicates a bug of Velox, while the throw of VELOX_USER_CHECK indicates that the exception is caused by user.

Delimiter size should not empty -> Non-empty delimiter is expected.

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,226 @@
+    const char* start = input.data();
+    re2::StringPiece subMatches[1];
+    while (re->Match(
+        re2String, pos, input.size(), RE2::Anchor::UNANCHORED, subMatches, 1)) {
```
Please explain how we match re2String against delim. Some example:

> Matches a regular expression against a portion of the input string, starting from 'pos' to the end of the input string. The match is not anchored, which means it can start at any position in the string. If a match is found, the matched portion of the string is stored in 'subMatches'. The '1' indicates that we are only interested in the first match found.

rui-mo reviewed (commented):
Could you run the fuzzer test for 5 minute using a command like below?
```
./velox/expression/fuzzer/spark_expression_fuzzer_test \
                 --seed ${RANDOM} \
                 --duration_sec 300 \
                 --minloglevel=0 \
                 --stderrthreshold=2 \
                 --only split
```

gaoyangxiaozhu commented:
> Could you run the fuzzer test for 5 minute using a command like below?
> 
> ```
> ./velox/expression/fuzzer/spark_expression_fuzzer_test \
>                  --seed ${RANDOM} \
>                  --duration_sec 300 \
>                  --minloglevel=0 \
>                  --stderrthreshold=2 \
>                  --only split
> ```

thanks @rui-mo , update a new commit to fix one regex match issue if the regex is not empty but has empty size match, and also locally run fuzz test 5 minutes and it passed 

```
./velox/expression/fuzzer/spark_expression_fuzzer_test --seed 1 --enable_variadic_signatures --velox_fuzzer_enable_complex_types --duration_sec 300 --minloglevel=0 --stderrthreshold=2 --only split
```
![image](https://github.com/facebookincubator/velox/assets/9858245/db2aaabf-658f-4a83-ab03-1d44b92eae63)

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,231 @@
+  // array can be of any size.
+  void split(
+      StringView input,
+      const StringView delim,
```
nit: For `StringView` and `const StringView`, can we use the same style?

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,231 @@
+    const char* pos = begin;
+    while (pos < end && pos < limit + begin) {
+      arrayWriter.add_item().setNoCopy(StringView(pos, 1));
+      pos += 1;
```
I wonder if this works for unicode characters which could occupy more than 1 byte. Could you confirm and add some tests? Perhaps use `!utf_cont` to only split for the first byte.

gaoyangxiaozhu replied:
This function not work for unicode (non utf-8) characters, what do you means `Perhaps use !utf_cont to only split for the first byt` ? @rui-mo

rui-mo replied:
We could use `utf_cont` to check if a byte is continuous byte in a character. If no, the byte should belong to a new character and we can split it out.

rui-mo reviewed (commented):
Thank you for iterating.

rui-mo commented on velox/functions/sparksql/tests/SplitFunctionsTest.cpp:
```diff
@@ -14,96 +14,382 @@
-      const std::vector<std::optional<std::string>>& input,
-      std::optional<char> pattern,
-      const std::vector<std::optional<std::vector<std::string>>>& output);
+  /// Method runs the given split function, f.e. split(C0, C1), where C0 is the
```
nit: use // for multiple-line comment of non-public API.

rui-mo commented on velox/functions/sparksql/tests/SplitFunctionsTest.cpp:
```diff
@@ -14,96 +14,382 @@
+  /// For expected result vectors, for some combinations of input encodings, we
+  /// need to massage the expected vector.
+  /// Const we wrap in const, dictionary we wrap in dictionary and the reast
+  /// leave 'as is'. In the end we flatten.
```
Ditto.

rui-mo commented on velox/functions/sparksql/tests/SplitFunctionsTest.cpp:
```diff
@@ -14,96 +14,382 @@
-    return input[row] ? StringView(*input[row]) : StringView();
+/**
+ * Test split vector function on vectors with different encodings.
+ */
```
nit: use //

gaoyangxiaozhu commented:
I also run again locall for fuzz test 5 minutes and works fine

gaoyangxiaozhu commented:
@rui-mo any new comments

rui-mo reviewed (commented):
Thanks. Looks good to me overall. @PHILO-HE Do you have any comment?

gaoyangxiaozhu commented on velox/functions/sparksql/tests/SplitFunctionsTest.cpp:
```diff
@@ -14,96 +14,420 @@
+      {"o", "n", "e", ",", ",", ",", "f", "o", "u", "r", ","},
+      {""},
+  });
+  assertEqualVectors(expected, run(inputStrings, delim, "split(C0, C1)"));
```
add `no limit` , `limit 0`, `limit -1`, `high limit`, `small limit` test case for empty delimiter case.

gaoyangxiaozhu commented on velox/docs/functions/spark/string.rst:
```diff
@@ -195,22 +195,22 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
-        SELECT split('oneAtwoBthreeC', '[ABC]', 0); -- ["one", "two", "three", ""]
         SELECT split('oneAtwoBthreeC', '[ABC]', 2); -- ["one","twoBthreeC"]
+        SELECT split('one', '1'); -- ["one"]
+        SELECT split('abcd', ''); -- ["a", "b", "c", "d"]
```
removed the white-spaces to aligin the style

rui-mo commented on velox/docs/functions/spark/string.rst:
```diff
@@ -195,22 +195,24 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+    contain all input beyond the last matched regex. When ``limit`` <= 0, ``regex`` will be applied as many
+    times as possible, and the resulting array can be of any size. When ``delimiter`` is empty, if ``limit``
+    is smaller than the size of ``string``, the resulting array only contains ``limit`` number of single characters
+    splitting from ``string``, if limit is not provided or is larger than the size of string, the resulting array 
```
nit: limit -> \``limit\``, string -> \``string\``

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,246 @@
+      // empty matches, instead of inserting an empty character into the result
+      // array, always split the character at the current 'pos' of the string
+      // and put it into the result array. Thus, the result array for the above
+      // example would be ["a", "b", "c"].
```
I find Spark's result contains an empty tail string, could you confirm?

> spark.sql("select split('abc', 'd|')").show(false)
+------------------+
|split(abc, d|, -1)|
+------------------+
|[a, b, c, ]       |
+------------------+

gaoyangxiaozhu replied:
yes, my mistake, actually we do contains an empty tail string and also have test case to cover https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/functions/sparksql/tests/SplitTest.cpp#L277-L282, just correct the comment.

jinchengchenghh commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,245 @@
+      const auto* rawStrings = strings->data<StringView>();
+      const auto delim = delims->valueAt<StringView>(0);
+      int32_t limit = std::numeric_limits<int32_t>::max();
+      if (limits) {
```
Can we extract the `limit` computation to a new inline function? I wonder if there is really user case that the `limit` is not constant, all the test in spark the limit is constant.

gaoyangxiaozhu replied:
done to extract `limit` to a inline function, and from my point, even `limit` in most cases is `constant`, we still need to keep
non-constant path due to spark don't declare it only support constant `limit`.

jinchengchenghh replied:
Before this patch, the second argument `pattern` is constant, I would suggest the new added `limit` also be constant.
CC @rui-mo

rui-mo replied:
Spark allows non-constant input for them, so I assume we need to support it. 

> Before this patch, the second argument pattern is constant

This is a limitation mentioned in https://github.com/facebookincubator/velox/issues/4066 which shall be addressed.

gaoyangxiaozhu commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,245 @@
-      resultWriter.commit();
-    });
+    // Fast path for (flat, const, const).
+    if (strings->isIdentityMapping() and delims->isConstantMapping() and
```
i am ok for both, but presto split here using `and`

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,244 @@
-        arrayWriter.add_item().setNoCopy(StringView(pos, delim - pos));
-        pos = delim + 1; // Skip past delim.
-      } while (delim != end);
+    auto getLimit = [&](int32_t row) -> int32_t {
```
nit: int32_t -> vector_size_t for row

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,244 @@
+      }
+
+      // Hit an empty match for the specific case when the delimiter has '|' as
+      // a suffix. For example, if the string is 'abc' and the delimiter is 'd|'
```
Is it a general handling for empty match or just for the case when `delimiter has '|' as suffix` as you mentioned? By reading the code, I assume it is a general handling, then we could make the comment more concise. Below is some example:

> When hitting an empty match, split the character at the current 'pos' and put it into the result array, followed by an empty tail string at last, e.g., the result array for split('abc', 'd|') is ["a", "b", "c", ""].

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,240 @@
 
```
nit: perhaps rename this file as Split.cpp

rui-mo commented on velox/functions/sparksql/SplitFunctions.cpp:
```diff
@@ -16,124 +16,240 @@
+    auto getLimit = [&](vector_size_t row) -> int32_t {
+      int32_t limit = std::numeric_limits<int32_t>::max();
+      if (limits) {
+        const auto limitValue = limits->valueAt<vector_size_t>(row);
```
use `valueAt<int32_t>` because limit is of int32_t type. We just changed to use vector_size_t for the index.

gaoyangxiaozhu replied:
sure thing, my mistake

gaoyangxiaozhu commented:
kindly ping @rui-mo for any new comments, thank you!

rui-mo reviewed (approved):
Thanks. Will mark it if no other comments from @PHILO-HE @jinchengchenghh.

gaoyangxiaozhu commented:
> Thanks. Will mark it if no other comments from @PHILO-HE @jinchengchenghh.

cool! thank you @rui-mo, then @PHILO-HE / @jinchengchenghh  please help update if you have any new comments

jinchengchenghh commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,452 @@
+          CppToType<StringView>::create(),
+          numRows,
+          std::make_unique<SimpleVectorLoader>(funcCreateFlatStrings));
+    } else if (isDictionary(encodingStrings)) {
```
Do we need to test the lazy and dictionary vector? This tests the framework, not the function itself. @rui-mo @PHILO-HE

gaoyangxiaozhu replied:
that's because we have fast path for different encoding here https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/functions/sparksql/Split.cpp#L64-L66 , otherwise we don't need this part tests.

jinchengchenghh commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,452 @@
+          numRows, [&](vector_size_t row) { return StringView{input[row]}; });
+    };
+
+    auto funcCreateFlatDelims = [&](RowSet /*rows*/) {
```
funcCreateFlatDelims -> createFlatDelims

jinchengchenghh commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,452 @@
+          numRows, [&](vector_size_t row) { return StringView{delim}; });
+    };
+
+    auto funcCreateFlatLimits = [&](RowSet /*rows*/) {
```
ditto

jinchengchenghh commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,452 @@
+          numRows, [&](vector_size_t row) { return limit.value(); });
+    };
+
+    auto funcReverseIndices = [&](vector_size_t row) {
```
ditto

PHILO-HE reviewed (commented):
Some comments. Thanks!

gaoyangxiaozhu commented on velox/functions/sparksql/Split.cpp:
```diff
@@ -0,0 +1,259 @@
+  if (inputArgs.size() > 2) {
+    VELOX_USER_CHECK(
+        inputArgs[2].type->kind() == TypeKind::INTEGER,
+        "The third argument should be of integer type.");
```
I actually have same thought, but i remember per rui's ask. let wait @rui-mo 's input.

rui-mo replied:
@PHILO-HE @gaoyangxiaozhu Thanks. I agree. This is not needed for this function. Perhaps remove this method and declare it as vector function instead of **stateful** vector function as below.

```
VELOX_DECLARE_VECTOR_FUNCTION(
    udf_regexp_split,
    signatures(),
    std::make_unique<Split>());
```

gaoyangxiaozhu commented on velox/functions/sparksql/Split.cpp:
```diff
@@ -0,0 +1,259 @@
+    };
+
+    // Fast path for (flat, const, const).
+    if (strings->isIdentityMapping() and delims->isConstantMapping() and
```
hey @PHILO-HE so you means this part optimization actually is not needed ? https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/functions/sparksql/Split.cpp#L67

![Uploading image.png…]()

gaoyangxiaozhu commented:
please help review and let's make sure if can make this PR ready this week @PHILO-HE / @jinchengchenghh , thank you!

jinchengchenghh reviewed (approved):
Only one @PHILO-HE 's question, I think we should check string is flat for strings->data() optimization, but we also have a else branch that not check string only check delimiter and limit is constant, because both dictionary vector and lazy vector is common. We should not let it move to non-constant delimiter and non-constant limit branch.

gaoyangxiaozhu commented:
> Only one @PHILO-HE 's question, I think we should check string is flat for strings->data() optimization, but we also have a else branch that not check string only check delimiter and limit is constant, because both dictionary vector and lazy vector is common. We should not let it move to non-constant delimiter and non-constant limit branch.

i got your point , but it is actually optional only for another optmization. Let me do in following patch.

gaoyangxiaozhu commented:
@PHILO-HE any new comments ?

rui-mo commented:
> but it is actually optional only for another optmization. Let me do in following patch.

@gaoyangxiaozhu It make sense to me to add current fast path. Perhaps we can take it as a follow-up to use a benchmark to determine if further optimization is needed.

PHILO-HE reviewed (approved):
I have no other comment.

gaoyangxiaozhu commented:
> > but it is actually optional only for another optmization. Let me do in following patch.
> 
> @gaoyangxiaozhu It make sense to me to add current fast path. Perhaps we can take it as a follow-up to use a benchmark to determine if further optimization is needed.
> 
> Would you address this comment [#10248 (comment)](https://github.com/facebookincubator/velox/pull/10248#discussion_r1677375306) ?

done

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -212,22 +212,24 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
-
-    Splits ``string`` on ``delimiter`` and returns an array. ::
+.. spark:function:: split(string, delimiter[, limit]) -> array(string)
+    Splits ``string`` around occurrences that match ``delimiter`` and returns an array with a length of
```
@gaoyangxiaozhu Would you generate the docs and verify they render nicely?

It might be helpful to add some examples, especially for cases when delimiter is empty.

gaoyangxiaozhu replied:
there are example for empty delimiter - https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/docs/functions/spark/string.rst?plain=1#L231-L234

mbasmanova commented on velox/functions/sparksql/Split.cpp:
```diff
@@ -0,0 +1,239 @@
+ public:
+  Split() {}
+
+  void apply(
```
Is there any particular reason this function is not implemented using Simple Function API?

gaoyangxiaozhu replied:
not, so you want to convert it to simple function ?

mbasmanova replied:
If there is no reason for this function to be written as Vector function, then it is better to write it as a Simple function. Simple functions are easier to implement and maintain. Thanks.

gaoyangxiaozhu replied:
@mbasmanova thanks,  I will do the convert, but please document this part if it is mandatory to write it as a simple function instead of using a vector function unless there is a performance improvement or another significant reason, otherwise it is easy to increase the burden on developers since we may wrongly implement vector way.
 Or if there already has a doc addressing this part please share with me. Thanks.

rui-mo replied:
> If there is no reason for this function to be written as Vector function, then it is better to write it as a Simple function. Simple functions are easier to implement and maintain.

Adding some documentation for this point: https://github.com/facebookincubator/velox/pull/10528. @mbasmanova Do you think it makes sense? Thanks.

mbasmanova replied:
@rui-mo Thank you, Rui. Accepted the PR with a small comment.

gaoyangxiaozhu replied:
@mbasmanova done to convert to simple. @rui-mo / @mbasmanova / @PHILO-HE to help review.

mbasmanova commented on velox/functions/sparksql/Split.cpp:
```diff
@@ -0,0 +1,239 @@
+// By default, limit is -1.
+class Split final : public exec::VectorFunction {
+ public:
+  Split() {}
```
not needed

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1431,4 +1433,143 @@ struct LevenshteinDistanceFunction {
+template <typename T>
+struct SplitFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
```
An initialize method could be added to handle constant delimiter and limit.

gaoyangxiaozhu replied:
any example code can reference ? @rui-mo , thanks

rui-mo replied:
I understand. I agree that initialize cannot bring much benefits as the processing is row by row.

gaoyangxiaozhu replied:
ping @rui-mo for above help ask.

rui-mo replied:
Some documentation on constant inputs: https://github.com/facebookincubator/velox/blob/main/velox/docs/develop/scalar-functions.rst#access-to-session-properties-and-constant-inputs.

gaoyangxiaozhu replied:
i saw this part doc but still not sure how to leverage do optimize @rui-mo

rui-mo replied:
I double-checked and found initialize cannot bring much benefits for this function, seeing https://github.com/facebookincubator/velox/pull/10248#discussion_r1689372828.

rui-mo commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -995,5 +996,298 @@ TEST_F(StringTest, trim) {
+  });
+  assertEqualVectors(expected, result);
+
+  limit = -1;
```
nit: perhaps add some comments to explain the purpose of tests for readability.

rui-mo reviewed (commented):
Thanks!

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1431,4 +1433,143 @@ struct LevenshteinDistanceFunction {
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& delimiter) {
+    doCall(result, input, delimiter, std::numeric_limits<int32_t>::max());
```
nit: perhaps use INT32_MAX.

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1431,4 +1433,143 @@ struct LevenshteinDistanceFunction {
+        result,
+        input,
+        delimiter,
+        limit > 0 ? limit : std::numeric_limits<int32_t>::max());
```
Ditto.

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1431,4 +1433,143 @@ struct LevenshteinDistanceFunction {
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // add them to the elements vector, until we reach the end of the
+    // string or the limit.
```
Comment suggestion:
> Splits input string using the delimiter and adds the cutting-off pieces to elements vector until the string's end or the limit is reached.

rui-mo commented on velox/functions/sparksql/String.h:
```diff
@@ -1431,4 +1433,143 @@ struct LevenshteinDistanceFunction {
+      // When hitting an empty match, split the character at the current 'pos'
+      // of the input string and put it into the result array, followed by an
+      // empty tail string at last, e.g., the result array for split('abc','d|')
+      // is ["a", "b", "c",""].
```
nit: add one space before `""` or remove all spaces between strings.

rui-mo commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -995,5 +996,298 @@ TEST_F(StringTest, trim) {
+      {""}, // The whole string is empty
+  };
+
+  auto createStrings = [&](RowSet /*rows*/) {
```
rows are not in use, can we create the string vector without it like below?
> makeFlatVector<StringView>({"ALGERIA", "ARGENTINA"})

gaoyangxiaozhu replied:
can't, we need leverage `row` to init with `data[row]`

rui-mo commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -995,5 +996,298 @@ TEST_F(StringTest, trim) {
+
+  auto createStrings = [&](RowSet /*rows*/) {
+    return makeFlatVector<StringView>(
+        numRows, [&](vector_size_t row) { return StringView{data[row]}; });
```
Why do we create vectors with the same elements? I assume we don't need to test duplicate values.

gaoyangxiaozhu replied:
not same right ? `data[row]` @rui-mo

gaoyangxiaozhu replied:
I do refactor and put the test in seperated test file `SplitTest.cpp`

jinchengchenghh commented on velox/functions/sparksql/tests/StringTest.cpp:
```diff
@@ -995,5 +996,298 @@ TEST_F(StringTest, trim) {
+  };
+
+  // No limit.
+  auto strings = createStrings({});
```
Add a new Helper function in class StringTest
`testSplit(const VectorPtr& expected, const VectorPtr& input, std::optional<>limit)` will help code clean

jinchengchenghh replied:
And function testSplit(const VectorPtr& expected, const VectorPtr& input)

rui-mo commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,261 @@
+      const std::vector<std::string>& input,
+      std::string delim,
+      size_t numRows,
+      std::optional<int32_t> limit) {
```
nit: adjust the order as (input, delim, limit, numRows, expected) for readability.

rui-mo commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,261 @@
+      std::optional<int32_t> limit) {
+    auto strings = makeFlatVector(input);
+    auto delims = makeFlatVector<StringView>(
+        numRows, [&](vector_size_t row) { return StringView{delim}; });
```
Is it better to accept a vector of delimiters and a vector of limits instead?

gaoyangxiaozhu replied:
thought before, I think current is better, l'd like to keep current way, otherwise we need construct delimiters vector and limits vector again and again for each test cases outside using `delim` and `limit` when the  `delim` or `limit ` values changed.

rui-mo commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,159 @@
+/// limit is an integer which controls the number of times the regex is applied.
+/// By default, limit is -1.
+template <typename T>
+struct SplitFunction {
```
nit: perhaps rename it as Split to align with the header name.

rui-mo reviewed (commented):
Thanks. Just several nits.

gaoyangxiaozhu commented:
@mbasmanova for any new comments ?

jinchengchenghh commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,159 @@
+      StringView input,
+      StringView delim,
+      int32_t limit) const {
+    VELOX_USER_CHECK(!delim.empty(), "Non-empty delimiter is expected");
```
VELOX_DCHECK, This logic is guaranteed by you.
```
spark-sql (default)> select split("a,b,c", "");
["a",",","b",",","c"]
Time taken: 0.093 seconds, Fetched 1 row(s)
```
The delimiter can be empty. Curious why this cannot be tested. Do I miss something?

gaoyangxiaozhu replied:
please  check `splitEmptyDelimiter` ~~

jinchengchenghh replied:
So it should `VELOX_DCHECK`, user won't meet this exception

jinchengchenghh commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,261 @@
+    auto delims = makeFlatVector<StringView>(
+        numRows, [&](vector_size_t row) { return StringView{delim}; });
+    if (limit.has_value()) {
+      auto limits = makeFlatVector<int32_t>(
```
makeConstant

jinchengchenghh commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,261 @@
+      size_t numRows,
+      const std::vector<std::vector<std::string>>& expected) {
+    auto strings = makeFlatVector(input);
+    auto delims = makeFlatVector<StringView>(
```
make constant

gaoyangxiaozhu commented:
@mbasmanova ping again

gaoyangxiaozhu commented:
this PR been there 2 month, would be speed up the review process, thank you

jinchengchenghh commented:
LGTM Thanks!

mbasmanova commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,159 @@
+ */
+#pragma once
+
+#include "velox/functions/prestosql/Utf8Utils.h"
```
sparksql should not depend on prestosql. Perhaps, move this to lib.

mbasmanova commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,159 @@
+/// Splits string on delimiter and returns an array of size at most limit.
+/// delimiter is a string representing regular expression.
+/// limit is an integer which controls the number of times the regex is applied.
+/// By default, limit is -1.
```
perhaps, explain that -1 means 'unlimited'.

mbasmanova commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,159 @@
+    int32_t count = 0;
+    while (pos < end && count < limit) {
+      auto charLength = tryGetCharLength(start + pos, end - pos);
+      VELOX_DCHECK_GT(charLength, 0);
```
Why is this a debug-only check? This condition may happen in release build as well if input is not a well-formed UTF-8 and it needs to be handled properly. CC: @rui-mo

gaoyangxiaozhu replied:
good catch!  Let me fix this to handle string which contains invalid-UTF-8 char. @rui-mo

jinchengchenghh replied:
Please also test  invalid-UTF-8 char

mbasmanova commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,159 @@
+  void split(
+      out_type<Array<Varchar>>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& delim,
```
do not abbreviate: delim -> delimiter

mbasmanova commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,159 @@
+      // empty tail string at last, e.g., the result array for split('abc','d|')
+      // is ["a","b","c",""].
+      if (size == 0) {
+        offset += tryGetCharLength(start + pos, end - pos);
```
same comment

mbasmanova reviewed (commented):
@gaoyangxiaozhu Some comments.

rui-mo commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,168 @@
+    while (pos < end && count < limit) {
+      auto charLength = tryGetCharLength(start + pos, end - pos);
+      if (charLength <= 0) {
+        // invalid UTF-8 character.
```
nit: capitalize the first letter for comments.

Comment suggestion:

> Invalid UTF-8 character is treated as single character.

rui-mo commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,168 @@
+      if (size == 0) {
+        auto charLength = tryGetCharLength(start + pos, end - pos);
+        if (charLength <= 0) {
+          // invalid UTF-8 character.
```
Ditto.

rui-mo reviewed (commented):
Added two nits. Thanks.

rui-mo commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,274 @@
+      auto input = makeRowVector({strings, delims, limits});
+      auto result = evaluate("split(c0, c1, c2)", input);
+      auto expectedResult = makeArrayVector(expected);
+      assertEqualVectors(expectedResult, result);
```
nit: expectedResult and assertEqualVectors can be reused for the two branches.

gaoyangxiaozhu commented:
ping @mbasmanova

mbasmanova commented on velox/docs/functions/spark/string.rst:
```diff
@@ -236,22 +236,26 @@ Unless specified otherwise, all functions return NULL if at least one of the arg
+    is smaller than the size of ``string``, the resulting array only contains ``limit`` number of single characters
+    splitting from ``string``, if ``limit`` is not provided or is larger than the size of ``string``, the resulting 
+    array contains all the single characters of ``string`` and does not include an empty tail character.
+    The split function align with vanilla spark 3.4+ split function. ::
```
> The split function align with vanilla spark 3.4+ split function.

Wondering why do we need this comment? Isn't it true for all functions that they are expected to match Spark 3.4+ behavior?

gaoyangxiaozhu replied:
just in case people see different behavior happen for spark version below 3.4 as 3.2/3.3

mbasmanova replied:
What about other functions? Do they implement Spark 3.2/3.3 semantics? CC: @rui-mo

mbasmanova replied:
@gaoyangxiaozhu @rui-mo gentle ping

rui-mo replied:
@mbasmanova For the split function, the behavior for empty delimiter was changed since Spark 3.4 with commit https://github.com/apache/spark/commit/247306c0fe8cfda17c4182ce2338f7312e3aa897.
When implementing functions in Velox, if there is semantic difference among Spark versions, I assume we need to follow the latest one. Does it makes sense? Thanks.

mbasmanova replied:
@rui-mo Rui, thank you for clarifying. I assume this is a general policy that applies to all functions. If so, it would be helpful to document it in https://facebookincubator.github.io/velox/spark_functions.html and clarify which Spark version we match and whether we match ANSI mode on or off.

Furthermore, assuming Spark doesn't guarantee backwards compatibility across versions, we would have to pick a specific version we match and cannot say "latest" or 3.4+. If Spark changes behavior in "latest" version and we change to match, existing users will start seeing different behavior. Is this the case?

rui-mo replied:
@mbasmanova Thanks for the pointer. I'd like to document it.
>  If Spark changes behavior in "latest" version and we change to match, existing users will start seeing different behavior. Is this the case?

I assume it is the case, and it makes sense to me to pick a specific version.

rui-mo replied:
Opened https://github.com/facebookincubator/velox/pull/10677 for the documentation.

@gaoyangxiaozhu I assume below statement and `Since Spark 3.4` for `splitEmptyDelimiter` function could be removed.

> The split function align with vanilla spark 3.4+ split function.

mbasmanova commented on velox/functions/prestosql/tests/Utf8Test.cpp:
```diff
@@ -15,7 +15,7 @@
  */
 
 #include <gtest/gtest.h>
-#include "velox/functions/prestosql/Utf8Utils.h"
```
Is this a test for Utf8Utils that were moved to lib? If so, the test needs to move to lib/tests.

The logic in Utf8Utils is Presto-specific. Does it happen to match Spark's behavior? Would you clarify? CC: @rui-mo

gaoyangxiaozhu replied:
Spark doesn't handle for invalid UTF-8 and wide-character,  spark only guarantee work well for char with size at most 2 bytes. It should be fine here to leverage Utf8Utils.

mbasmanova replied:
> spark only guarantee work well for char with size at most 2 bytes 

UDF-8 utilities we have support characters up to 4 bytes long. Are you saying Spark only allows 2-byte characters? If so, do we expect different behavior for strings with 3-4 byte characters? CC: @rui-mo

gaoyangxiaozhu replied:
sorry @mbasmanova , i mix other string function with this split function. Dobule checked for split function, spark convert it to `UTF8String` class to do split action, it can handle `wide character` well,  looks the only case spark can't handle well is the case if input string has `invalid UTF-8` and delimiter is empty.

![image](https://github.com/user-attachments/assets/2e4cb38e-f6f4-4bed-bbec-cd3c087f8c34)

Checking code here - https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L581

rui-mo replied:
Looks like inputs are assumed to be valid UTF8 characters as @mbasmanova  mentioned at https://github.com/facebookincubator/velox/pull/10264#discussion_r1684218012.

gaoyangxiaozhu replied:
@mbasmanova

mbasmanova replied:
@rui-mo @gaoyangxiaozhu @PHILO-HE To avoid confusion going forward, should we update documentation in https://facebookincubator.github.io/velox/functions/spark/string.html to clearly state that input strings are expected to be vaild UTF-8 and that behavior is undefined if they are not?

rui-mo replied:
Opened https://github.com/facebookincubator/velox/pull/10682. Thanks.

mbasmanova commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,168 @@
+    while (pos < end && count < limit) {
+      auto charLength = tryGetCharLength(start + pos, end - pos);
+      if (charLength <= 0) {
+        // Invalid UTF-8 character is treated as single character.
```
Is this behavior aligned with Spark? CC: @rui-mo

gaoyangxiaozhu replied:
as i shared, spark doesn't handle for invalid UTF-8 and wide-character, it only guarantee work well for char with size at most 2 bytes (UTF-16), that means if input has invalid UTF-8 character as `\xE0`, it would treat it `\`, `x`, `E`, `0`. Here we treat a signer character.

mbasmanova reviewed (commented):
@gaoyangxiaozhu Looks good % a few remaining questions / comments. Please, address @rui-mo 's comments.

gaoyangxiaozhu commented:
kindly ping @mbasmanova

mbasmanova commented on velox/functions/sparksql/Split.h:
```diff
@@ -0,0 +1,168 @@
+      auto charLength = tryGetCharLength(start + pos, end - pos);
+      if (charLength <= 0) {
+        // Invalid UTF-8 character is treated as single character.
+        charLength = 1;
```
Given the following behavior of tryGetCharLength, would it make sense to set charLength to -charLength ? If not, why? 

> A negative return value means
> /// the UTF-8 sequence at the position is invalid, and the length of the invalid
> /// sequence is the absolute value of the result.

gaoyangxiaozhu replied:
thanks @mbasmanova , i didn't before noticed this comment, but it's cool if the absolute value of the return value is the invalid character sequence length, then we can levearge that, instead of processing the invalid character sequence one by one.

After that, the continuous invalid character sequence for example `\xED\xA0` would be treat one not two `\xED`, `\xA0`.

mbasmanova commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,282 @@
+ protected:
+  void testSplit(
+      const std::vector<std::string>& input,
+      std::string delim,
```
pass std::string as const &

mbasmanova commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,282 @@
+      size_t numRows,
+      const std::vector<std::vector<std::string>>& expected) {
+    auto strings = makeFlatVector(input);
+    auto delims = makeConstant<StringView>(StringView{delim}, numRows);
```
Does makeConstant(delim, numRows) not work?

gaoyangxiaozhu replied:
not, but `makeConstant(StringView{delim}, numRows)` works.

mbasmanova commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,282 @@
+};
+
+TEST_F(SplitTest, basic) {
+  auto limit = -1;
```
Wondering if the test will be more readable if this variable is removed and its usage is replaced with hard-coded values.

```
  // Negative limit.
  testSplit(input, delim, -1, numRows, expected);

  // Limit greater than the input string size.
  testSplit(input, delim, 100, numRows, expected);
```

gaoyangxiaozhu replied:
got! Remove all `limit` and `delim` variable and replace with hardcode values.

mbasmanova commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,282 @@
+  // Non-ascii delimiter.
+  delim = "లేదా";
+  input = {
+      {"синяя сливаలేదా赤いトマトలేదా黃苹果లేదాbrown pear"},
```
Curious, what do the Chinese characters say? Would you share translation into English? (Somehow, got worried about what we are hard-coding... want to make sure these words are appropriate.)

gaoyangxiaozhu replied:
here `黃苹果` means `yellow apple`, and `赤いトマト` is japanese which means `red tomato`.

mbasmanova replied:
Nice, thank you for explaining.

mbasmanova commented on velox/functions/sparksql/tests/SplitTest.cpp:
```diff
@@ -0,0 +1,282 @@
+  testSplit(input, delim, limit, numRows, expected);
+
+  // Cover case that delimiter not exists in input string.
+  delim = "A";
```
Ditto. Might be more readable to hard-coded delim:

```
testSplit(input, "A", std::nullopt, numRows, expected);
```

gaoyangxiaozhu replied:
done

mbasmanova reviewed (commented):
@gaoyangxiaozhu Overall looks good. A few remaining nits.

mbasmanova reviewed (approved):
Thanks!

FelixYBW commented:
@xiaoxmeng Can you merge the PR?
